torchrl/objectives/crossq.py

```diff
@@ -0,0 +1,622 @@
+        entropy = -metadata_actor["log_prob"]
+        out = {
+            "loss_actor": loss_actor,
+            "loss_qvalue": loss_qvalue,
+            "loss_alpha": loss_alpha,
+            "alpha": self._alpha,
+            "entropy": entropy.detach().mean(),
+        }
+        td_out = TensorDict(out, [])
+        td_out = td_out.named_apply(
+            lambda name, value: (
+                _reduce(value, reduction=self.reduction)
+                if name.startswith("loss_")
+                else value
+            ),
+            batch_size=[],
+        )
+        return td_out
+
+    @property
+    @_cache_values
+    def _cached_detached_qvalue_params(self):
+        return self.qvalue_network_params.detach()
+
+    def _actor_loss(
+        self, tensordict: TensorDictBase
+    ) -> Tuple[Tensor, Dict[str, Tensor]]:
+        with set_exploration_type(
+            ExplorationType.RANDOM
+        ), self.actor_network_params.to_module(self.actor_network):
+            dist = self.actor_network.get_dist(tensordict)
+            a_reparm = dist.rsample()
+        log_prob = dist.log_prob(a_reparm)
+
+        td_q = tensordict.select(*self.qvalue_network.in_keys, strict=False)
+        td_q.set(self.tensor_keys.action, a_reparm)
+        td_q = self._vmap_qnetworkN0(
+            td_q,
+            self._cached_detached_qvalue_params,  # should we clone?
+        )
+        min_q_logprob = (
+            td_q.get(self.tensor_keys.state_action_value).min(0)[0].squeeze(-1)
+        )
+
+        if log_prob.shape != min_q_logprob.shape:
+            raise RuntimeError(
+                f"Losses shape mismatch: {log_prob.shape} and {min_q_logprob.shape}"
+            )
+
+        return self._alpha * log_prob - min_q_logprob, {"log_prob": log_prob.detach()}
+
+    def _qvalue_loss(
+        self, tensordict: TensorDictBase
+    ) -> Tuple[Tensor, Dict[str, Tensor]]:
+
+        # # compute next action
+        with torch.no_grad():
+            with set_exploration_type(
+                ExplorationType.RANDOM
+            ), self.actor_network_params.to_module(self.actor_network):
+                next_tensordict = tensordict.get("next").clone(False)
+                next_dist = self.actor_network.get_dist(next_tensordict)
+                next_action = next_dist.sample()
+                next_tensordict.set(self.tensor_keys.action, next_action)
+                next_sample_log_prob = next_dist.log_prob(next_action)
+
+        # next_state_action_value = self._vmap_qnetworkN0(
+        #     next_tensordict.select(*self.qvalue_network.in_keys, strict=False),
+        #     self.qvalue_network_params,
+        # ).get(self.tensor_keys.state_action_value)
+
+        # current_state_action_value = self._vmap_qnetworkN0(
+        #     tensordict.select(*self.qvalue_network.in_keys, strict=False),
+        #     self.qvalue_network_params,
+        # ).get(self.tensor_keys.state_action_value)
+
+        combined = torch.cat(
```

Comment:
Solved the previous issue but the sequential forward pass for current state and next state values is still faster than the combined. Cat and splitting might be slowing down the computation.

Reply:
Let's keep the commented vmaps, we can make them run faster eventually